cpp/include/cugraph/mtmg/resource_manager.hpp

```diff
@@ -0,0 +1,241 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cugraph/mtmg/handle.hpp>
+#include <cugraph/mtmg/instance_manager.hpp>
+#include <cugraph/partition_manager.hpp>
+
+#include <raft/comms/std_comms.hpp>
+
+#include <rmm/cuda_device.hpp>
+#include <rmm/exec_policy.hpp>
+#include <rmm/mr/device/cuda_memory_resource.hpp>
+#include <rmm/mr/device/pool_memory_resource.hpp>
+
+#include <execution>
+
+namespace cugraph {
+namespace mtmg {
+
+/**
+ * @brief Class for managing local and remote GPU resources for use in
+ *   multi-threaded multi-GPU interface.
+ *
+ * Each process in a multi-GPU configuration should have an instance of this
+ * class.  The resource manager object should be configured by calling
+ * register_local_gpu (or register_remote_gpu once we support a multi-node
+ * configuration) to allocate resources that can be used in the mtmg space.
+ *
+ * When we want to execute some graph computations, we need to create an instance for execution.
+ * Based on how big a subset of the desired compute resources is desired, we can allocate some
+ * number of GPUs to the problem (up to the total set of managed resources).
+ *
+ * The returned instance can be used to create a graph, execute one or more algorithms, etc.  Once
+ * we are done the caller can delete the instance.
+ *
+ * At the moment, the caller is assumed to be responsible for scheduling use of the resources.
+ *
+ * For our first release, we will only consider a single node multi-GPU configuration, so the remote
+ * GPU methods are currently disabled via ifdef.
+ */
+class resource_manager_t {
+ public:
+  resource_manager_t() {}
+
+  /**
+   * @brief add a local GPU to the resource manager.
+   *
+   * @param rank       The rank to assign to the local GPU
+   * @param device_id  The device_id corresponding to this rank
+   */
+  void register_local_gpu(int rank, rmm::cuda_device_id device_id)
+  {
+    std::lock_guard<std::mutex> lock(lock_);
+
+    CUGRAPH_EXPECTS(local_rank_map_.find(rank) == local_rank_map_.end(),
+                    "cannot register same rank multiple times");
+
+    int num_gpus_this_node;
+    RAFT_CUDA_TRY(cudaGetDeviceCount(&num_gpus_this_node));
+
+    CUGRAPH_EXPECTS((device_id.value() >= 0) && (device_id.value() < num_gpus_this_node),
+                    "device id out of range");
+
+    local_rank_map_.insert(std::pair(rank, device_id));
+
+    cudaSetDevice(device_id.value());
+
+    // FIXME: There is a bug in the cuda_memory_resource that results in a Hang.
+    //   using the pool resource as a work-around.
+    //
+    // Ultimately there should be some RMM parameters passed into this function
+    // (or the constructor of the object) to configure this behavior
```

Comment:
We can set `NCCL_LAUNCH_MODE=GROUP` to workaround this problem.

Reply:
And we may need to use this with the pool allocator as well assuming the pool allocator will use cudaMalloc/cudaFree for bulk allocation. Just the odd for hang is much lower with the pool allocator but it could be harder to reproduce and debug once it happens.

Reply:
I didn't know this is deprecated, better ask to the NCCL folks about a proper longer term solution to work around this problem.

Reply:
Sent a message to NCCL folks